src_analysis/fuzzer_dependencies.py

```diff
@@ -0,0 +1,185 @@
+
+2. import blah  # Relative-import (against style guide anyway).
+
+This case is not supported because the dependency will not be recognized:
+from fuzzers.afl.fuzzer import build
+
+"""
+import importlib
+import inspect
+import os
+import types
+from typing import Dict, List, Set
+import sys
+
+from common import fuzzer_utils
+
+# The max depth of dependencies _get_python_dependencies will search.
+PY_DEPENDENCIES_MAX_DEPTH = 10
+
+# A cache of Python dependencies for modules. Keys are strings of module paths.
+# Values are sets of module paths.
+PY_DEPENDENCIES_CACHE: Dict[str, Set[str]] = {}
+
+# Cache these so we don't need to do it every time we call
+# get_underlying_fuzzer.
+FUZZER_CONFIGS = fuzzer_utils.get_fuzzer_configs()
+FUZZER_NAMES_TO_UNDERLYING = {
+    fuzzer_utils.get_fuzzer_from_config(config): config['fuzzer']
+    for config in FUZZER_CONFIGS
+}
+
+
+def _get_fuzzer_module_name(fuzzer: str) -> str:
+    """Returns the name of the fuzzer.py module of |fuzzer|. Assumes |fuzzer| is
+    an underlying fuzzer."""
+    return 'fuzzers.{}.fuzzer'.format(fuzzer)
+
+
+def is_builtin_module(module: types.ModuleType) -> bool:
+    """Returns True if |module| is a python builtin module."""
+    return module.__name__ in sys.builtin_module_names
+
+
+def is_fuzzers_subpath(path: str) -> bool:
+    """Returns True if path is a subpath of the fuzzers/ directory."""
+    common_path = os.path.commonpath([path, fuzzer_utils.FUZZERS_DIR])
+    return common_path == fuzzer_utils.FUZZERS_DIR
+
+
+def is_fuzzers_submodule(module) -> bool:
+    """Returns True if |module| is a submodule of the fuzzers module."""
+    if is_builtin_module(module):
+        # builtin modules such as "time" don't have files so attempts to get
+        # their files fail. Check for these and bail early.
+        return False
+
+    # If a module imports `fuzzers` we can't handle it, let the TypeError get
+    # thrown so we fail loudly.
+    module_path = inspect.getfile(module)
+    return is_fuzzers_subpath(module_path)
+
+
+def get_fuzzer_dependencies(fuzzer: str) -> Set[str]:
+    """Returns the list of files in fuzzbench that |fuzzer| depends on. This
+    includes dockerfiles used to build |fuzzer|, and the python files it uses to
+    build and run fuzz targets."""
+    # TODO(metzman): Write a test for this that uses fakefs to enforce that
+    # every fuzzer can successfully run with only the dependencies this function
+    # finds.
+    initial_fuzzer = fuzzer
+    fuzzer = get_underlying_fuzzer(fuzzer)
+    fuzzer_directory = fuzzer_utils.FuzzerDirectory(fuzzer)
+
+    if initial_fuzzer == fuzzer:
+        dependencies = set()
+    else:
+        # If fuzzer's base fuzzer is different, fuzzer is a variant, which
+        # means changes to variants.yaml can affect it.
+        dependencies = {fuzzer_directory.variants_yaml}
+
+    # Don't use modulefinder for python dependencies since it doesn't work
```

Comment:
Modulefinder did not work here because the modules do not contain __init__.py files. We have three ways of dealing with this:
1. The route I took, use something other than modulefinder. It probably doesn't handle edge cases as well (such as importing functions, rather than modules, which the style guide doesn't allow anyway).
2. Forcing every module in fuzzers to have an __init__.py.
3. Hacking modulefinder to work anyway. I did this in a few lines of code by using fakefs and adding a fake __init__.py. But this seems too. hacky

Reply:
I personally like 1, especially if we can get it to handle `import fuzzers.afl.fuzzer` since that is really the only use case that isn't against the style guide that actually works but src_analysis module doesn't support (I don't think it's that hard to special case these kinds of imports). But it might not be worth the complexity so 2 is also a reasonable option (at the cost of minor integration inconvenience).

Reply:
1 sounds good, no need to force people to add __init__.py, feels crappy.

Reply:
ACK